### ISSUE TITLE
Can't import GLSL shaders in headless mode

### ISSUE THREAD

**Issue by TCROC** [godotengine/godot#94734]
### Tested versions

4.3.beta.mono.custom_build.ba7c5b0c8

### System information

Godot v4.3.beta.mono (ba7c5b0c8) - Pop!_OS 22.04 LTS - Wayland - Vulkan (Forward+) - dedicated Intel(R) Arc(tm) A750 Graphics (DG2) - AMD Ryzen 7 5800X 8-Core Processor (16 Threads)

### Issue description

Godot currently doesn't support importing glsl shaders in a headless context.  This creates a problem for continuous integration / pipelines.  @clayjohn worded it perfectly in the rocket chat so I'll just quote it here:

> This is a very valuable workflow when you are targeting multiple platforms and becomes super important when working with a team. Basically everyone on the team can work on the platform of their choice, then the CI makes an export build for all target platforms and then you can test release builds on all platforms

> Since CI doesn't have a GPU, you have to be able to import all resources, then export the game without access to a GPU. Ultimately, the game will be played on a device with a GPU, but the import/export process happens without one

This appears to be the PR that disabled glsl imports in headless mode: https://github.com/godotengine/godot/pull/92539

I tried simply reverting that PR and interestingly, it does seem to import the glsl shader.  However, it crashes in the exported project when that shader is loaded if it wasn't previously loaded in a non-headless context.

### Steps to reproduce

1. Clone the proton scatter repo or use the zipped mrp (which is also just the proton scatter repo): https://github.com/HungryProton/scatter.git
2. Open in godot
3. Notice that the trees are loading in the showcase scene.
4. Create an export preset
5. Delete the ``.godot`` folder in the project directory so there aren't any cached imports.
6. Run the command: ``godot --headless --path path/to/project --import``
7. Notice the errors regarding failed glsl imports.
8. Run the command: ``godot --headless --path path/to/project --export-debug PresetName path/to/export/game.exe``
9. Notice the errors regarding glsl
10. Start the exported game and notice that the trees no longer load

### Minimal reproduction project (MRP)

[scatter.zip](https://github.com/user-attachments/files/16377132/scatter.zip)

https://github.com/HungryProton/scatter.git

**Comment by TCROC** [comment]
@Repiteo this issue shouldn't be closed. The PRs that went through were for separate issues. I believe they may have even referenced the fact that they don't resolve this issue. I can see how an automated system may have incorrectly attached this issue to one of those PRs 😅

**Comment by TCROC** [comment]
@akien-mga @AThousandShips This issue should still be open.

Thanks everyone :)

**Comment by AThousandShips** [comment]
The reason this was closed is you wrote "when issue gets resolved: #94734", you linked it yourself, make sure you pay attention to the special labels (and pay attention to the side bar where they are listed)

**Comment by TCROC** [comment]
Ah I see. I will pay attention to it going into the future. Thanks for the heads up :)

**Comment by kisg** [comment]
I think this is a critical issue for any non-trivial project that uses CI export process and has shaders. Unless this issue is fixed, this limitation of the '--headless' switch should be clearly documented.

I would consider removing the dummy renderer/shader storage in favor of including a headless display server or using LibGodot's embedded display server with a headless configuration. In GPU-less environments, like a CI server, a software-only driver like Mesa's llvmpipe would be more than enough because the export process doesn't do any actual rendering, just imports and exports. This solution would reduce the codebase and the testing/maintenance needs.

An alternative would be to fix the dummy renderer/shader storage to handle the import/export of shaders properly. In this case, automated tests should be included to avoid the regression of the headless feature.

**Comment by clayjohn** [comment]
@kisg this only impacts RDShaders, it doesn't impact GDShaders.

**Comment by kisg** [comment]
@clayjohn So GDShader export should work with --headless? Because it didn't work for me.

**Comment by clayjohn** [comment]
> [@clayjohn](https://github.com/clayjohn) So GDShader export should work with --headless? Because it didn't work for me.

Yep, I've used it quite a bit. In 4.3 global uniforms and instance uniforms weren't working, but even they are supported now in 4.4. If you are having problems with GDShaders, please make a bug report with an MRP

**Comment by TCROC** [comment]
> [@TCROC](https://github.com/TCROC) [#102552](https://github.com/godotengine/godot/pull/102552) made some architectural changes that should allow importing GLSL shaders when in headless mode. The PR is still in draft, but please test it out once it is closer to being ready just to confirm that it does indeed fix this bug (in addition to its intended purpose)

Perfect!  I look forward to testing this out!  Please feel free to @ me whenever this is ready! :)

**Comment by clayjohn** [comment]
Reopening since I don't think this was actually fixed by the shader baker. The comment was "Should fix: xxx" and I think Github interpreted that wrongly

**Comment by beicause** [comment]
I tested the MRP in 4.5 on Linux and this seems to be fixed.
The GLSL files are successfully imported without crash in headless mode, and the exported game runs normally. I also enabled Shader Baker during export and no issues were found.

I notice the changes in #92539 is reverted in #102552. 

Godot v4.5.beta - EndeavourOS SMP PREEMPT_DYNAMIC Thu, 19 Jun 2025 14:41:19 +0000 on Wayland - X11 display driver, Multi-window, 1 monitor - Vulkan (Forward+) - integrated AMD Radeon Graphics (RADV RENOIR) - AMD Ryzen 7 4800U with Radeon Graphics (16 threads) - 14.98 GiB memory

**Comment by TCROC** [comment]
@clayjohn @beicause I can confirm that this is fixed! :)  It can't bake in headless tho.  But that is a separate issue.  For this particular issue of not being able to import, it is fixed :)
